src/library/blas3/clsparse-spm-spm.cpp

```diff
@@ -0,0 +1,1009 @@
+            else if (j == 127)
+                run_status = copy_Ct_to_C_Loop( 256, counter, j, counter_one[j], csrValC, csrRowPtrC, csrColIndC, csrValCt, csrRowPtrCt, csrColIndCt, queue_one, control);
+
+            if (run_status != CL_SUCCESS)
+            {
+                return clsparseInvalidKernelExecution;
+            }
+        }
+    }
+    
+    return clsparseSuccess;
+
+}
+ 
+
+ CLSPARSE_EXPORT clsparseStatus
+        clsparseScsrSpGemm( 
+        const clsparseCsrMatrix* sparseMatA,
+        const clsparseCsrMatrix* sparseMatB,
+              clsparseCsrMatrix* sparseMatC,
+        const clsparseControl control )
+{
+    cl_int run_status;
+
+    if (!clsparseInitialized)
+    {
+       return clsparseNotInitialized;
+    }
+
+    if (control == nullptr)
+    {
+       return clsparseInvalidControlObject;
+    }
+
+    const clsparseCsrMatrixPrivate* matA = static_cast<const clsparseCsrMatrixPrivate*>(sparseMatA);
+    const clsparseCsrMatrixPrivate* matB = static_cast<const clsparseCsrMatrixPrivate*>(sparseMatB);
+    clsparseCsrMatrixPrivate* matC = static_cast<clsparseCsrMatrixPrivate*>(sparseMatC);
+
+    int m  = matA->num_rows;
+    int k1 = matA->num_cols;
+    int k2 = matB->num_rows;
+    int n  = matB->num_cols;
+    int nnzA = matA->num_nonzeros;
+    int nnzB = matB->num_nonzeros;
+    
+    if(k1 != k2)
+    {
+        std::cerr << "A.n and B.m don't match!" << std::endl; 
+        return clsparseInvalidKernelExecution;
+    }  
+    
+    cl_mem csrRowPtrA = matA->rowOffsets;
+    cl_mem csrColIndA = matA->colIndices;
+    cl_mem csrValA    = matA->values;
+    cl_mem csrRowPtrB = matB->rowOffsets;
+    cl_mem csrColIndB = matB->colIndices;
+    cl_mem csrValB    = matB->values;
+    
+    cl::Context cxt = control->getContext();
+    
+    matC->rowOffsets = ::clCreateBuffer( cxt(), CL_MEM_READ_WRITE, (m + 1) * sizeof( cl_int ), NULL, &run_status );
+    
+    int pattern = 0;
+    clEnqueueFillBuffer(control->queue(), matC->rowOffsets, &pattern, sizeof(cl_int), 0, (m + 1)*sizeof(cl_int), 0, NULL, NULL);
+                        
+    cl_mem csrRowPtrC = matC->rowOffsets;
+    int *csrRowPtrC_h = (int *) malloc((m + 1) * sizeof(int));
+    memset(csrRowPtrC_h, 0, (m + 1) * sizeof(int));
+    CHECKMAL(csrRowPtrC_h, "csrRowPtrC_h");
+
+    cl_mem csrRowPtrCt_d = ::clCreateBuffer( cxt(), CL_MEM_READ_WRITE, (m + 1) * sizeof( cl_int ), NULL, &run_status );	   
+    clEnqueueFillBuffer(control->queue(), csrRowPtrCt_d, &pattern, sizeof(cl_int), 0, (m + 1)*sizeof(cl_int), 0, NULL, NULL);
+    
+    int *csrRowPtrCt_h = (int *) malloc((m + 1) * sizeof(int));
+    memset(csrRowPtrCt_h, 0, (m + 1) * sizeof(int));
+    
+    // STAGE 1
+    compute_nnzCt(m, csrRowPtrA, csrColIndA, csrRowPtrB, csrColIndB, csrRowPtrCt_d, control);
+    
+    // statistics
+    int *counter = (int *)malloc(NUM_SEGMENTS * sizeof(int));
```

Comment:
Maybe we can use std::array or std::vector for handling this parameters? This will eliminate use of CHECKMAL macro.